packages/salesforce-adapter/src/filters/custom_objects_instances.ts

```diff
@@ -0,0 +1,134 @@
+    if (curr.done) return undefined
+    if (_.isArray(curr.value) && curr.value.length > 0) {
+      res.push(...transformer(curr.value))
+    }
+    return next()
+  }
+  await next()
+  return res
+}
+
+const getObjectInstances = async (
+  client: SalesforceClient,
+  object: ObjectType
+): Promise<Array<InstanceElement>> => {
+  const recordsToInstances = (records: Values[]): InstanceElement[] => {
+    const recordToInstance = (record: Values): InstanceElement => {
+      const getInstancePath = (instanceName: string): string[] => {
+        const objectNamespace = getNamespace(object)
+        if (objectNamespace) {
+          return [SALESFORCE, INSTALLED_PACKAGES_PATH, objectNamespace,
+            RECORDS_PATH, object.elemID.typeName, instanceName]
+        }
+        return [SALESFORCE, RECORDS_PATH, object.elemID.typeName, instanceName]
+      }
+
+      return new InstanceElement(
+        // TODO: Handle elemID with additional logic
+        new ElemID(SALESFORCE, record.Id).name,
+        object,
+        record,
+        getInstancePath(record.Id),
+      )
+    }
+
+    const doesObjectIncludeNameField = (): boolean =>
+      _.some(Object.values(object.fields), (f => (isObjectType(f.type)
+      && f.type.elemID.isEqual(Types.compoundDataTypes.Name.elemID))))
+
+    // Name sub-fields are returned at top level -> move them to the nameField
+    const transformNameValues = (): Values[] => {
+      const nameSubFields = Object.keys(Types.compoundDataTypes.Name.fields)
+      const nameFieldName = Object.keys(_.pickBy(object.fields, f =>
+        (isObjectType(f.type)
+          // We assume there's only one Name field
+          && f.type.elemID.isEqual(Types.compoundDataTypes.Name.elemID))))[0]
+      return records.map(value => {
+        const valuesWithoutNameFields = _.pickBy(
+          value,
+          (_value, keyName) => !nameSubFields.includes(keyName)
+        )
+        return Object.assign(valuesWithoutNameFields, {
+          [nameFieldName]: _.pickBy(
+            value,
+            (_value, keyName) => nameSubFields.includes(keyName)
+          ),
+        })
+      })
+    }
+
+    const instanceValues = doesObjectIncludeNameField() ? transformNameValues() : records
+
+    return instanceValues.map(recordToInstance)
+  }
+  const queryString = buildQueryString(object)
+  const recordsIterable = await client.queryAll(queryString)
+  return consumeAsyncIterable(recordsIterable, recordsToInstances)
+}
+
+const getObjectTypesInstances = async (
+  client: SalesforceClient,
+  objects: ObjectType[]
+): Promise<InstanceElement[]> =>
+  (_.flatten(await Promise.all(objects.map(o => getObjectInstances(client, o)))))
+
+const filterObjectTypes = (elements: Element[], namespaces: string[]): ObjectType[] =>
+  (elements.filter((e: Element) =>
+    isCustomObject(e) && isObjectType(e)
+    && (!_.isUndefined(getNamespace(e))
+      && namespaces.includes(getNamespace(e) as string))) as ObjectType[])
```

Comment:
nit: we should avoid casting when possible
```suggestion
const filterObjectTypes = (elements: Element[], namespaces: string[]): ObjectType[] => (
  elements
    .filter(isObjectType)
    .filter(isCustomObject)
    .filter(e => {
      const namespace = getNamespace(e)
      return namespace !== undefined && namespaces.includes(namespace)
    })
)
```